### PR TITLE
crimson/osd: drop a foreign-copy to shard-0 for every pg operation

### PR DESCRIPTION
**Pull request by cyx1231st**
By using ConnectionRef before pg submission, and after that, change to use ConnectionXcoreRef.

The intent is to drop the foreign copy of the connection to shard 0 at pg submission time. This should remove two pairs of crosscore communications in shard 0 for each I/O, one for connection-ref foreign copy, another for connection-ref destruction.

---

After this PR, in a local test with 32 cores, 1 replication, 4K random rados write, cyanstore, the outstanding tasks in shard 0 is reduced by ~10x, reactor utilizations are all close to 100, which looks that the imbalance is no longer significant in this case, and IOPS is about ~10% better.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/osd/pg_shard_manager.h [hunk 1]
```diff
@@ -190,17 +193,17 @@ class PGShardManager {
...
+    auto &crosscore_ordering = get_osd_priv(
+        &opref.get_local_connection()).crosscore_ordering;
     auto cc_seq = crosscore_ordering.prepare_submit(core);
     auto &logger = crimson::get_logger(ceph_subsys_osd);
     logger.debug("{}: send {} to the remote pg core {}",
                  opref, cc_seq, core);
     return opref.get_handle().complete(
-    ).then([&opref, this] {
-      get_local_state().registry.remove_from_registry(opref);
-      return opref.prepare_remote_submission();
```
Comment (cyx1231st) [inline src/crimson/osd/pg_shard_manager.h:201]: The most essential change of this PR is that the foreign copy to shard 0 here is no longer necessary.

**Review by athanatos — APPROVED** [review]

**Review by liu-chunmei — APPROVED** [review]

**Comment by cyx1231st** [comment]
rebased

**Comment by cyx1231st** [comment]
jenkins test windows

**Comment by cyx1231st** [comment]
rebased onto the approved cleanup https://github.com/ceph/ceph/pull/55080

File: src/crimson/osd/osd_operations/logmissing_request_reply.cc [hunk 2]
```diff
@@ -21,7 +21,7 @@ namespace crimson::osd {
 LogMissingRequestReply::LogMissingRequestReply(
   crimson::net::ConnectionRef&& conn,
   Ref<MOSDPGUpdateLogMissingReply> &&req)
-  : conn{std::move(conn)},
+  : l_conn{std::move(l_conn)},
```
Comment (cyx1231st) [inline src/crimson/osd/osd_operations/logmissing_request_reply.cc:24]: This should really be:
`: l_conn{std::move(conn)},`

Comment (cyx1231st) [reply src/crimson/osd/osd_operations/logmissing_request_reply.cc]: fixed.

**Comment by cyx1231st** [comment]
rebased

**Comment by athanatos** [comment]
> Added a potential fix here: [12e4686](https://github.com/ceph/ceph/commit/12e4686aee4cae72453c09f13ec877e9b23a5a65)

That's correct -- scrub events should be ordered with peering requests rather than client requets.

**Comment by cyx1231st** [comment]
https://pulpito.ceph.com/yingxin-2024-01-23_01:28:42-crimson-rados-wip-yingxin-crimson-save-conn-foreign-copy-9-distro-default-smithi/

7527503: https://tracker.ceph.com/issues/63647
7527511: failed, but not sure about the cause

@Matan-B @athanatos how do you think about the above test results ?

**Comment by Matan-B** [comment]
> https://pulpito.ceph.com/yingxin-2024-01-23_01:28:42-crimson-rados-wip-yingxin-crimson-save-conn-foreign-copy-9-distro-default-smithi/
> 
> 7527503: https://tracker.ceph.com/issues/63647 7527511: failed, but not sure about the cause
> 
> @Matan-B @athanatos how do you think about the above test results ?

@cyx1231st, second failure failed because of:
```
2024-01-23T07:32:35.736 INFO:tasks.workunit.client.0.smithi059.stdout:[  FAILED  ] 1 test, listed below:
2024-01-23T07:32:35.736 INFO:tasks.workunit.client.0.smithi059.stdout:[  FAILED  ] TestLibRBD.TestEncryptionLUKS2
```
Which is known: https://tracker.ceph.com/issues/63844

Tests LGTM!

**Comment by cyx1231st** [comment]
If no further concerns, I'll proceed to merge https://github.com/ceph/ceph/pull/55080, then https://github.com/ceph/ceph/pull/54896

**Pull request merged**